test/e2e/tas/mpijob_test.go

```diff
@@ -0,0 +1,177 @@
+			)
+
+			numPods := launcherReplicas + workerReplicas
+
+			mpijob := testingmpijob.MakeMPIJob("ranks-mpi", ns.Name).
+				Queue(localQueue.Name).
+				MPIJobReplicaSpecs(
+					testingmpijob.MPIJobReplicaSpecRequirement{
+						Image:         util.E2eTestSleepImage,
+						Args:          []string{"60s"},
+						ReplicaType:   kfmpi.MPIReplicaTypeLauncher,
+						ReplicaCount:  launcherReplicas,
+						RestartPolicy: corev1.RestartPolicyOnFailure,
+						Annotations: map[string]string{
+							kueuealpha.PodSetPreferredTopologyAnnotation: topologyLevelRack,
+						},
+					},
+					testingmpijob.MPIJobReplicaSpecRequirement{
+						Image:         util.E2eTestSleepImage,
+						Args:          []string{"60s"},
+						ReplicaType:   kfmpi.MPIReplicaTypeWorker,
+						ReplicaCount:  workerReplicas,
+						RestartPolicy: corev1.RestartPolicyOnFailure,
+						Annotations: map[string]string{
+							kueuealpha.PodSetPreferredTopologyAnnotation: topologyLevelBlock,
+						},
+					},
+				).
+				Request(kfmpi.MPIReplicaTypeLauncher, extraResource, "1").
+				Limit(kfmpi.MPIReplicaTypeLauncher, extraResource, "1").
+				Request(kfmpi.MPIReplicaTypeWorker, extraResource, "1").
+				Limit(kfmpi.MPIReplicaTypeWorker, extraResource, "1").
+				Obj()
+			gomega.Expect(k8sClient.Create(ctx, mpijob)).Should(gomega.Succeed())
+
+			ginkgo.By("MPIJob is unsuspended", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(mpijob), mpijob)).To(gomega.Succeed())
+					g.Expect(mpijob.Spec.RunPolicy.Suspend).Should(gomega.Equal(ptr.To(false)))
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			pods := &corev1.PodList{}
+			ginkgo.By("ensure all pods are created", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.List(ctx, pods, client.InNamespace(ns.Name))).To(gomega.Succeed())
+					g.Expect(pods.Items).Should(gomega.HaveLen(numPods))
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("ensure all pods are scheduled", func() {
+				listOpts := &client.ListOptions{
+					FieldSelector: fields.OneTermNotEqualSelector("spec.nodeName", ""),
+				}
+				gomega.Eventually(func(g gomega.Gomega) {
+					g.Expect(k8sClient.List(ctx, pods, client.InNamespace(ns.Name), listOpts)).To(gomega.Succeed())
+					g.Expect(pods.Items).Should(gomega.HaveLen(numPods))
+				}, util.LongTimeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("verify the assignment of pods are as expected with rank-based ordering", func() {
+				gomega.Expect(k8sClient.List(ctx, pods, client.InNamespace(ns.Name))).To(gomega.Succeed())
+				gotAssignment := readRankAssignmentsFromMPIJobPods(pods.Items)
+				wantAssignment := map[string]string{
+					"worker/0": "kind-worker",
+					"worker/1": "kind-worker2",
+					"worker/2": "kind-worker3",
+				}
+				gomega.Expect(wantAssignment).Should(gomega.BeComparableTo(gotAssignment))
```

Comment:
My initial thought was that launcher would get kind-worker (0), and the workers will get `kind-worker2, kind-worker3, kind-worker4`, because the launcher should be processed first as it is first on the list of PodSets: https://github.com/kubernetes-sigs/kueue/blob/15e00ea0f8371e4fef6e8f67f8fc1f4d39bad6c9/pkg/controller/jobs/mpijob/mpijob_controller.go#L204-L213. Maybe I'm missing something.

Please:
- assert on the location of the launcher
- add unit test with launcher and worker and check which is first processed by TAS

Reply:
For launcher it gets random one. Not sure how to check it on e2e...

Reply:
Launcher don't have index.

Reply:
I synced with @mbobrovskyi and we decided to use cpu for launcher, this way it can be placed on the same node as one of the workers (worker0) in this case. This also corresponds better to the intended use.